src/configurable_adam.py

```diff
@@ -0,0 +1,214 @@
+    :note: if any of these devices are CPU, they will be prefetched for optimizer step using pinned memory
+    :param compute_dtype: dtype for optimizer step computation; defaults to param dtype
+    """
+
+    def __init__(
+            self,
+            params: Iterable[Union[torch.Tensor, dict]],
+            lr: float = 1e-3,
+            betas: Tuple[float] = (0.9, 0.999),
+            eps: float = 1e-6,
+            weight_decay: float = 0,
+            debias: Optional[bool] = None,
+            amsgrad: bool = False,
+            lamb: bool = False,
+            clamp_value: Optional[float] = None,
+            compute_dtype: Optional[torch.dtype] = None,
+            exp_avg_dtype: Optional[torch.dtype] = None,
+            exp_avg_sq_dtype: Optional[torch.dtype] = None,
+            v_hat_max_dtype: Optional[torch.dtype] = None,
+            exp_avg_device: torch.device = None,
+            exp_avg_sq_device: torch.device = None,
+            v_hat_max_device: torch.device = None,
+    ) -> None:
+        defaults = dict(
+            lr=lr, betas=betas, eps=eps, weight_decay=weight_decay, debias=debias, amsgrad=amsgrad, lamb=lamb,
+            clamp_value=clamp_value, compute_dtype=compute_dtype, exp_avg_dtype=exp_avg_dtype,
+            exp_avg_sq_dtype=exp_avg_sq_dtype, v_hat_max_dtype=v_hat_max_dtype, exp_avg_device=exp_avg_device,
+            exp_avg_sq_device=exp_avg_sq_device, v_hat_max_device=v_hat_max_device,
+        )
+        super().__init__(params, defaults)
+
+    def _maybe_init_state(self, param: torch.Tensor, group: dict) -> dict:
+        state = self.state[param]
+        if "step" not in state:
+            state["step"] = 0
+        if group["betas"][0] != 0 and "exp_avg" not in state:
+            pin_memory = group["exp_avg_device"] == torch.device("cpu")
+            state["exp_avg"] = torch.zeros_like(
+                param, dtype=group['exp_avg_dtype'],
+                memory_format=torch.preserve_format,
+                device=group["exp_avg_device"],
+                pin_memory=pin_memory)
+        if group["betas"][1] not in (0, 1) and "exp_avg_sq" not in state:
+            pin_memory = group["exp_avg_sq_device"] == torch.device("cpu")
+            state["exp_avg_sq"] = torch.zeros_like(
+                param, dtype=group['exp_avg_sq_dtype'],
+                memory_format=torch.preserve_format,
+                device=group["exp_avg_sq_device"],
+                pin_memory=pin_memory)
+        if group["amsgrad"] and "v_hat_max" not in state:
+            pin_memory = group["v_hat_max_device"] == torch.device("cpu")
+            state["v_hat_max"] = torch.zeros_like(
+                param, dtype=group['v_hat_max_dtype'],
+                memory_format=torch.preserve_format,
+                device=group["v_hat_max_device"],
+                pin_memory=pin_memory)
+        return state
+
+    @torch.no_grad()
+    def step(self, closure: Optional[callable] = None):
```

Comment:
get_loss?

Reply:
can't change name for compatibility with torch optimizer